nixos/modules/virtualisation/virtualbox-guest.nix

```diff
@@ -67,27 +75,33 @@ in
         # Allow systemd dependencies on vboxguest.
         SUBSYSTEM=="misc", KERNEL=="vboxguest", TAG+="systemd"
       '';
-  } (mkIf cfg.x11 {
-    services.xserver.videoDrivers = [ "vmware" "virtualbox" "modesetting" ];
 
-    services.xserver.config =
-      ''
-        Section "InputDevice"
-          Identifier "VBoxMouse"
-          Driver "vboxmouse"
-        EndSection
-      '';
+    systemd.user.services.virtualboxClient = {
+      description = "VirtualBox Guest User Services";
 
-    services.xserver.serverLayoutSection =
-      ''
-        InputDevice "VBoxMouse"
-      '';
+      wantedBy = [ "graphical-session.target" ];
+      partOf = [ "graphical-session.target" ];
 
-    services.xserver.displayManager.sessionCommands =
-      ''
-        PATH=${makeBinPath [ pkgs.gnugrep pkgs.which pkgs.xorg.xorgserver.out ]}:$PATH \
-          ${kernel.virtualboxGuestAdditions}/bin/VBoxClient-all
-      '';
-  })]);
+      # The graphical session may not be ready when starting the service
+      # Hence, check if the DISPLAY env var is set, otherwise fail, wait and retry again
+      startLimitBurst = 20;
+
+      unitConfig.ConditionVirtualization = "oracle";
 
+      # Check if the display environment is ready, otherwise fail triggering a restart of the service
+      preStart = "${pkgs.bash}/bin/bash -c \"if [ -z $DISPLAY ]; then exit 1; fi\"";
+      serviceConfig = {
+        ExecStart = pkgs.writeShellScript "virtualboxGuestAdditions-wrapper" ''
+          ${optionalString (cfg.clipboard) "${kernel.virtualboxGuestAdditions}/bin/VBoxClient --clipboard"}
```

Comment:
nit: You can put `kernel.virtualboxGuestAdditions` in `systemd.user.services.virtualboxClient.path` to save you lots of typing here. Then you can just refer to `VBoxClient` directly.